eax/src/online.rs

```diff
@@ -0,0 +1,326 @@
+/// which is when the resulting tag is equal to the one created during encryption.
+///
+/// # Panic
+/// If the `Eax` value will not be consumed via [`finish`] the
+/// process will abort, when compiled with `std` feature enabled, to prevent
+/// against bugs related to decrypting data without verifying its authenticity.
+/// ## Example
+/// ```
+/// use eax::{Error, online::{Eax, Decrypt, Encrypt}};
+/// use aes::Aes256;
+/// use block_cipher::generic_array::GenericArray;
+///
+/// let key = GenericArray::from_slice(b"an example very very secret key.");
+///
+/// let nonce = GenericArray::from_slice(b"my unique nonces"); // 128-bits; unique per message
+///
+/// let assoc = b"my associated data";
+/// let plaintext = b"plaintext message";
+///
+/// let mut buffer: [u8; 17] = *plaintext;
+///
+/// // Encrypt a simple message
+/// let mut cipher = Eax::<Aes256, Encrypt>::with_key_and_nonce(key, nonce);
+/// cipher.update_assoc(&assoc[..]);
+/// cipher.encrypt(&mut buffer[..9]);
+/// cipher.encrypt(&mut buffer[9..]);
+/// let tag = cipher.finish();
+///
+/// assert_ne!(buffer, *plaintext);
+///
+/// let mut cloned = buffer;
+///
+/// // Now decrypt it, using the same key and nonce
+/// let mut cipher = Eax::<Aes256, Decrypt>::with_key_and_nonce(key, nonce);
+/// cipher.update_assoc(&assoc[..]);
+/// cipher.decrypt(&mut buffer[..5]);
+/// cipher.decrypt(&mut buffer[5..10]);
+/// cipher.decrypt(&mut buffer[10..]);
+/// let res = cipher.finish(&tag);
+///
+/// assert_eq!(res, Ok(()));
+/// assert_eq!(buffer, *plaintext);
+///
+/// // Decrypting the ciphertext with tampered associated data should fail
+/// let mut cipher = Eax::<Aes256, Decrypt>::with_key_and_nonce(key, nonce);
+///
+/// cipher.update_assoc(b"tampered");
+/// cipher.decrypt(&mut cloned);
+/// let res = cipher.finish(&tag);
+///
+/// assert_eq!(res, Err(Error));
+/// ```
+///
+/// [<sup>1</sup>]: https://en.wikipedia.org/wiki/Online_algorithm
+/// [`Eax`]: ../struct.Eax.html
+/// [`Decrypt`]: struct.Decrypt.html
+/// [`finish`]: #method.finish
+pub struct Eax<Cipher, Op>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+    Op: CipherOp,
+{
+    nonce: Nonce<Cipher::BlockSize>,
+    data: Cmac<Cipher>,
+    message: Cmac<Cipher>,
+    ctr: ctr::Ctr128<Cipher>,
+    /// Denotes whether this stream is used for encryption or decryption.
+    marker: PhantomData<Op>,
+    /// Verifies at run-time whether the type has been properly consumed via
+    /// `Eax::finish`, otherwise aborts.
+    bomb: DropBomb,
+}
+
+/// Runtime-enforced linear-ish type.
+///
+/// This type is useful to enforce that a value is correctly explicitly consumed.
+/// Otherwise, this will abort (only when `std` feature is enabled).
+struct DropBomb;
+#[cfg(feature = "std")]
+impl Drop for DropBomb {
+    fn drop(&mut self) {
+        std::eprintln!("Drop bomb says buh-bye");
+        std::process::abort();
```

Comment:
While I understand the purpose of this code, I am not really comfortable with it. I would prefer to mark "online AEAD" as a hazmat operation. We already (intentionally) expose a fair amount of operations, which could be dangerously misused without a proper knowledge, so I don't think we should add such fragile guard rails.

Reply:
I don't mind and would be fine with exposing the online set of operations as unsafe (in the general sense). Originally, I thought that the harder-to-misuse API approach would be preferred, so I erred on the safe side and wrote this, instead.

I'll remove the drop bomb code.